### DISCUSSION TITLE
Querying Mongo

### DISCUSSION THREAD

**Discussion by raghunami25** [Budibase/budibase#1886] [Help]
Hey There,
 I want to try querying external mongo DB, but I didn't find any documentation on it. I tried doing it by myself, but it didn't work and am not able to figure out the query. Can you please help me?

Thanks.

**Comment by shogunpurple** [comment]
Hi @raghunami25

- when you say it didn’t work, what exactly occurred? Which steps did you take? 
- Could you maybe show what your query/confíg looks like and we can help you fix it. 

Thanks

**Comment by raghunami25** [comment]
Hi @shogunpurple 
This is how i have done config for DB source.
[
![Screenshot 2021-07-01 at 1 43 26 AM](https://user-images.githubusercontent.com/25820652/124025919-a9e9c800-da0e-11eb-86bf-6574982a885e.png)
](url)

This is how I was trying to query DB. 
![Screenshot 2021-07-01 at 1 52 16 AM](https://user-images.githubusercontent.com/25820652/124026365-3a280d00-da0f-11eb-933c-66ad0f72882e.png)

Can you provide a document regarding syntax for mongo querying, I don't see it anywhere in the documentation.

**Reply by shogunpurple** [reply]
Ah @raghunami25 - what you put in the box there is what gets passed to mongos `find`.

Change `db.buyerinput.find({})` to `{}`

That will give you the effect you are looking for, provided that your connection parameters are correct.

**Reply by raghunami25** [reply]
Return with a error can not use a session that has ended.

**Comment by che-122** [comment]
I've also been testing out MongoDB

To get it to work, you need to specify the collection you are querying against.

- ConnectionString: <yourstring>
- Db: YourDB
- Collection: buyerinput

Then you need to use a different syntax, you need to use the syntax for Compass. So `db.mycollection.find({})` wont work, you need to work with json style syntax. On the docs site for MongoDB there's a dropdown where you can choose Compass as the syntax to see the examples in the correct format. 
https://docs.mongodb.com/manual/tutorial/query-documents/

With that said, I also find it a bit hard to work with mongodb on budibase, it doesn't work as I'd expect. For example needing to create a new datasource for each collection seems a bit weird. The only DB methods available are create and read.

**Reply by che-122** [reply]
I'd be up for adding some features/improve the mongodb experience if someone on the team points me in the right direction

**Reply by shogunpurple** [reply]
@faroutchris Thanks for taking the time. Yeah, I have to agree. Specifying a different datasource per collection is probably not the best approach, and it should be at the query level.

This is our mongoDB integration code: 
https://github.com/Budibase/budibase/blob/master/packages/server/src/integrations/mongodb.ts

I'd love to hear what kind of improvements you had in mind, and we can help you out with this.

Thanks!

**Reply by che-122** [reply]
Nice! I'm thinking that you'd move collection out of the base config object and specify it for each query instead.
So if I create a mongo datasource, I'd only specify ConnectionString and Db.
And for each query, I'd want to specify the table/collection.

I'd also expand the list of methods available to all the ones mongo supports:
- listCollections
- find/findOne/findOneAndUpdate
- count
- distinct
- aggregate
- insertOne/Many
- updateOne/Many
- deleteOne/Many

~~Is there a guide to set up a local dev environment?~~ Nevermind, found it.

**Reply by raghunami25** [reply]
> I've also been testing out MongoDB
> 
> To get it to work, you need to specify the collection you are querying against.
> 
> * ConnectionString:
> * Db: YourDB
> * Collection: buyerinput
> 
> Then you need to use a different syntax, you need to use the syntax for Compass. So `db.mycollection.find({})` wont work, you need to work with json style syntax. On the docs site for MongoDB there's a dropdown where you can choose Compass as the syntax to see the examples in the correct format.
> https://docs.mongodb.com/manual/tutorial/query-documents/
> 
> With that said, I also find it a bit hard to work with mongodb on budibase, it doesn't work as I'd expect. For example needing to create a new datasource for each collection seems a bit weird. The only DB methods available are create and read.

Thanks it's really helpful.

**Reply by raghunami25** [reply]
> @faroutchris Thanks for taking the time. Yeah, I have to agree. Specifying a different datasource per collection is probably not the best approach, and it should be at the query level.
> 
> This is our mongoDB integration code:
> https://github.com/Budibase/budibase/blob/master/packages/server/src/integrations/mongodb.ts
> 
> I'd love to hear what kind of improvements you had in mind, and we can help you out with this.
> 
> Thanks!

It would be really helpful.

**Reply by raghunami25** [reply]
Read functions don't work with all the collections. For some collections, it returns with an error - can not use a session that has ended.

**Reply by che-122** [reply]
@shogunpurple I'm having some trouble with the local environment. It seems that the frontend portion only worked for a short while. Without touching the frontend code it seems to get stuck on the init/mount of the layout.svelte and won't continue rendering. I've disabled cache, adblockers and so on in case some broswer extension interferes with the site.

I've also tried nuking and running the setup several times. Unsure what could be the cause.

```
// The error
VM732:1 Uncaught (in promise) SyntaxError: Unexpected token < in JSON at position 0

// Seems to catch on this line in _layout.svelte (onMount)
await admin.init()

// This is the JSON it tries to parse, which is just the HTML document
<!doctype html>
<html class="spectrum spectrum--medium spectrum--darkest" lang="en" dir="ltr">
<head>
<script type="module" src="/builder/@vite/client"></script>
	<meta charset='utf8'>
	<meta name='viewport' content='width=device-width'>
	<title>Budibase</title>
	<link rel='icon' href="/builder/src/favicon.ico">
	<link rel="preconnect" href="https://fonts.gstatic.com" />
	<link
		href="https://fonts.googleapis.com/css2?family=Source+Sans+Pro:wght@400;600;700&display=swap"
		rel="stylesheet"
	/>
</head>
<body id="app">
	<script type="module" src="/builder/src/main.js"></script>
</body>
</html>
```

**Reply by shogunpurple** [reply]
Hi @faroutchris, are you accessing the builder at localhost:10000?

If possible, it would also be good if you could look at the devtools network response to see if a specific error is being thrown from the backend. This would help a lot with debugging.

Also, just wanted to touch on this:

>     listCollections
>     find/findOne/findOneAndUpdate
>     count
>     distinct
>     aggregate
>     insertOne/Many
>     updateOne/Many
>     deleteOne/Many

Right now, the external data sources low level API sticks to the CRUD methods to keep things fairly simple. We would ideally want to have each of these methods remain, but the underlying configuration would allow users to acheive all of the above through fields/toggles etc.

**Reply by che-122** [reply]
@shogunpurple Yeah, I'm using the correct port. It's weird, I managed to get it set up the first time and run the project locally. Seems the services in docker are running and nothing weird in the terminal. I'll try a few things out and see what happens

**Reply by shogunpurple** [reply]
@faroutchris if you had a standard budibase instance already on your machine set up with the budibase CLI, there could be some clashing issues with the production containers being used. Make sure all of your budibase containers are stopped, then go through the process again

**Comment by jlamoreaux** [comment]
👋  Hi! Stumbled upon this, checking out Budibase as a possible replacement for our current tool. I was able to connect to MongoDB, but I haven't been able to figure out how to query using an ObjectId. I've tried the standards like `{$oid: [id]}`, `{objectId: [id]}` but nothing seems to work. Any advice?

**Reply by shogunpurple** [reply]
Hi @jlamoreaux - I was able to recreate this - it's because the `_id` field in your mongo document is an `ObjectId`, and `find` does not know how to parse this into an `ObjectId`.

the new mongo integration by @faroutchris has not yet been releaased, but will be this week. It contains the `findOne` method, which you need to use to search by ID and the mongo node driver handles the conversion from string to `ObjectId`. 

Right now, You can search on non-id fields, but I'm afraid you will have to wait for the release this week to do what you are looking for. 

Thanks!

**Reply by zyazhb** [reply]
@shogunpurple  
Is this feature aviliable now? 
I have tried to query by `_id`, but still not working.

**Reply by shogunpurple** [reply]
Hi @zyazhb - yep, this should be available now, with the `findOne` method included. Is it still not working for you? What does your query look like?

**Reply by zyazhb** [reply]
@shogunpurple I use it like 
```json
{
	"_id": "614c2cf4e14dfe71a3c8a998"
}
```
in query, but continue got Query Error: cannot convert undifiend or null object.

I also tried `_id 614c2cf4e14dfe71a3c8a998` in Parameters, it returned a incorrect result.
Im sure this document is avilable in my db.

**Reply by phuclb1nxt** [reply]
anyone try this? i still got same error with findOne

**Reply by phuclb1nxt** [reply]
@shogunpurple pls help. Thanks

**Reply by melohagan** [reply]
Hi @phuclb1nxt 

What error are you seeing? Also is your id a string, or is it an ObjectId? 
If it's an ObjectId you'll need the syntax: `"_id": "ObjectId('614c2cf4e14dfe71a3c8a998')"`

**Reply by phuclb1nxt** [reply]
oh , thank you. I solved it.

**Comment by Pgedela** [comment]
HI @shogunpurple 
I am currently rialing out BB and looks pretty awesome so far.
I am connecting to MongoDB, and following this thread, managed to get queries working.

Wondering, is there is a way to project the fields in mongo i.e. only select certain properties, 
similar to this https://docs.mongodb.com/manual/tutorial/project-fields-from-query-results/

Our collection is huge and dont need all of it.

Greatly appreciate your inputs.

Thanks !

**Reply by mjashanks** [reply]
Hi @Pgedela - we don't currently support projections with MongoDB. We don't have any immediate plans to update the MongoDb connector, but happy to accept contributions to it.

At the end of this month, we will be introducing "Transformers" - whereby you can write javascript inside Budibase, to run on your collections as they get returned from MongoDB. You could use this to achieve the same result. However, it will still pull the same dataset into Budibase before running that code, so it's not going to help performance - just tidy things up a bit.

**Reply by Pgedela** [reply]
got it @mjashanks 'Transformers' itslef is a good start to keep things clean !

**Comment by PhillipgDF** [comment]
Hi. I cant seem to get update function to work with my MongoDB/Budibase I can connect to Read but Update constantaly provides a cant convert and undefined or null to object error. Should I be referencing _id? I'll confess Im a newbie on this.

**Reply by zyazhb** [reply]
I was confused about that as well
https://github.com/Budibase/budibase/discussions/1886#discussioncomment-1777937

**Comment by Zonalds** [comment]
There really should be documentation on MongoDB usage. Half the time, we just have to guess and it makes the entire process difficult.

**Reply by mjashanks** [reply]
Agree - have tagged this as `documention`. We will be having a big push for this next month. Thanks for flagging.

**Comment by PhillipgDF** [comment]
For Mongo Queries - the Greater than $gt or $gte does not seem to function as it throws an error. What format is it expecting? Mongo Compass or?
